Posts/2023/01/Mastering-The-Steppable-Pipeline.md

```diff
@@ -0,0 +1,272 @@
+
+### Processing blocks
+
+As you might have noticed, some actions, like outputting a header, are only required only ones. As in the analogy with the assembly line, heating up a soldering gun is only required ones when the pipeline is started, cleaning up the station is only required when the pipeline is completed. Similar time consuming ("expensive") actions could be required for a cmdlet as e.g. open and closing a file. These actions are respectively defined in the `Begin` - and  `End` block of a cmdlet. The actual processing is defined in the `Process` block of cmdlet.
+A well defined pipeline PowerShell cmdlet might look like this:
+
+```PowerShell
+function MyCmdlet {
+    [CmdletBinding()] param(
+        [Parameter(ValueFromPipeLine = $True)][String] $InputString
+    )
+    Begin {
+        $Stream = [System.IO.StreamWriter]::new("$($Env:Temp)\My.Log")
+    }
+    Process {
+        $Stream.WriteLine($_)
+    }
+    End {
+        $Stream.Close()
+    }
+}
+```
+
+Using the above cmdlet example with the command like `1..9 | MyCmdlet`, the log file is *only opened once* at the start of the pipeline, than each item in the pipeline is process one-at-a-time and at the end the log file is closed (*once*).
+Note that when there are no `Begin`, `Process` and  `End` processing blocks defined in a function, the content of the function is assigned to the `End` block. See also: [about Functions Advanced Methods](https://learn.microsoft.com/powershell/module/microsoft.powershell.core/about/about_functions_advanced_methods)
+
+A similar pipeline can be created with the common [`Foreach-Object`](https://learn.microsoft.com/powershell/module/microsoft.powershell.core/foreach-object) cmdlet using the `-Begin`, `-Process` and `-End` parameters to define the corresponding process blocks:
+
+```PowerShell
+1..9 | Foreach-Object -Begin {
+    $Stream = [System.IO.StreamWriter]::new("$($Env:Temp)\My.Log")
+} -Process {
+    $Stream.WriteLine($_)
+} -End {
+    $Stream.Close()
+}
+```
+
+### Performance
+
+With the above explanation of the pipeline, you might understand now why you shouldn't wrap cmdlet pipeline in other pipeline, like:
+
+```PowerShell
+1..9 | ForEach-Object {
+    $_ | MyCmdlet
+}
+```
+
+Wrapping a cmdlet pipeline into another (`ForEach-Object`) pipeline is very expensive because you're also invoking the `begin` and `end` block of `MyCmdlet` which will open and close the concerned log file for each specific item instead of only once (at the beginning and the end of the pipeline). The performance degradation appears with native cmdlets along with `Export-Csv -append` and `Add-Content`. See also [PowerShell scripting performance considerations](https://learn.microsoft.com/powershell/scripting/dev-cross-plat/performance/script-authoring-considerations#avoid-wrapping-cmdlet-pipelines)
```

Comment:
```suggestion
Wrapping a cmdlet pipeline into another (`ForEach-Object`) pipeline is very expensive because you're also invoking the `begin` and `end` block of `MyCmdlet`. This will open and close the log file for each item instead of only once at the beginning and the end of the pipeline. The performance degradation can happend with any cmdlet that takes pipeline input. See also [PowerShell scripting performance considerations](https://learn.microsoft.com/powershell/scripting/dev-cross-plat/performance/script-authoring-considerations#avoid-wrapping-cmdlet-pipelines).
```